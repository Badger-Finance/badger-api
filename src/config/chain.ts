import { ethers } from 'ethers';
import { SettDefinition } from '../interface/Sett';
import { Token } from '../interface/Token';
import { Provider } from './constants';
import { ethSetts } from './setts/eth-setts';

export interface Chain {
  readonly name: string;
  readonly symbol: string;
  readonly chainId: string;
  readonly tokens: Token[];
  readonly setts: SettDefinition[];
  readonly provider: ethers.providers.JsonRpcProvider;
}

export class Ethereum implements Chain {
  public name = 'Ethereum';
  public symbol = 'eth';
  public chainId = '0x01';
  public tokens = [];
  public setts = ethSetts;
  public provider = new ethers.providers.JsonRpcProvider(Provider.Cloudflare);
}

export class BinanceSmartChain implements Chain {
  public name = 'BinanceSmartChain';
  public symbol = 'bsc';
  public chainId = '0x61';
  public tokens = [];
  public setts = [];
  public provider = new ethers.providers.JsonRpcProvider(Provider.Binance);
}

export const eth = new Ethereum();
export const bsc = new BinanceSmartChain();

export const supportedChains = [eth, bsc];

/**
 * Resolve a request query param chain lookup.
 * Failure to specify a chain, or requesting a non-supported
 * chain will default to Ethereum.
 * TODO: Should not supported throw a bad request instead?
 */
export const resolveChainQuery = (symbol: string): Chain => {
  if (!symbol) return eth;
  const chain = supportedChains.find((c) => c.symbol.toLowerCase() === symbol.toLowerCase());
  if (!chain) return eth;
  return chain;
};
