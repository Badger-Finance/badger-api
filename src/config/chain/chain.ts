import { BadRequest } from '@tsed/exceptions';
import { ethers } from 'ethers';
import { SettDefinition } from '../../interface/Sett';
import { Token } from '../../interface/Token';
import { Provider } from '../constants';
import { bscSetts } from './bsc';
import { ethSetts } from './eth';

export interface Chain {
	readonly name: string;
	readonly symbol: string;
	readonly chainId: string;
	readonly tokens: Token[];
	readonly setts: SettDefinition[];
	readonly provider: ethers.providers.JsonRpcProvider;
}

export class Ethereum implements Chain {
	public name = 'Ethereum';
	public symbol = 'eth';
	public chainId = '0x01';
	public tokens = [];
	public setts = ethSetts;
	public provider = new ethers.providers.JsonRpcProvider(Provider.Cloudflare);
}

export class BinanceSmartChain implements Chain {
	public name = 'BinanceSmartChain';
	public symbol = 'bsc';
	public chainId = '0x61';
	public tokens = [];
	public setts = bscSetts;
	public provider = new ethers.providers.JsonRpcProvider(Provider.Binance);
}

export const eth = new Ethereum();
export const bsc = new BinanceSmartChain();

export const supportedChains = [eth, bsc];

/**
 * Resolve a request query param chain lookup.
 * Failure to specify a chain, or requesting a non-supported
 * chain will default to Ethereum.
 */
export const resolveChainQuery = (symbol: string): Chain => {
	if (!symbol) return eth;
	const chain = supportedChains.find((c) => c.symbol.toLowerCase() === symbol.toLowerCase());
	if (!chain) throw new BadRequest('Invalid chain');
	return chain;
};
