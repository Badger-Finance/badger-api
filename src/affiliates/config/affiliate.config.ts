import { BadRequest } from '@tsed/exceptions';
import { Chain } from '../../chains/config/chain.config';
import { Protocol } from '../../config/constants';
import { SettAffiliateData } from '../../setts/interfaces/sett-affiliate-data.interface';
import { SettDefinition } from '../../setts/interfaces/sett-definition.interface';

type Affiliates = { [protocol: string]: Affiliate };

export abstract class Affiliate {
  private static affiliates: Affiliates = {};

  readonly name: string;

  constructor(name: string) {
    this.name = name;
  }

  static register(protocol: Protocol, affiliate: Affiliate): void {
    this.affiliates[protocol] = affiliate;
  }

  static getAffiliate(affiliateData: SettAffiliateData): Affiliate {
    const affiliate = this.affiliates[affiliateData.protocol];
    if (!affiliate) {
      throw new BadRequest(`${affiliate} is not a affiliate`);
    }
    return affiliate;
  }

  abstract getAffiliateVaultData(chain: Chain, sett: SettDefinition): Promise<SettAffiliateData>;
}
