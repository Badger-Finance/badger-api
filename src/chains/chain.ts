import { Arbitrum } from './config/arbitrum.config';
import { BinanceSmartChain } from './config/bsc.config';
import { Chain } from './config/chain.config';
import { Ethereum } from './config/eth.config';
import { Polygon } from './config/matic.config';

/**
 * Instantiate objects for registration.
 */
export const loadChains = (): Chain[] => {
  return [new Ethereum(), new BinanceSmartChain(), new Polygon(), new Arbitrum()];
};
