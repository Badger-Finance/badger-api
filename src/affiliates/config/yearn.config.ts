import { BigNumber, ethers } from 'ethers';
import { Chain } from '../../chains/config/chain.config';
import { guestListAbi } from '../../config/abi/guest-list.abi';
import { yearnAffiliateVaultWrapperAbi } from '../../config/abi/yearn-affiliate-vault-wrapper.abi';
import { Protocol } from '../../config/constants';
import { SettAffiliateData } from '../../setts/interfaces/sett-affiliate-data.interface';
import { SettDefinition } from '../../setts/interfaces/sett-definition.interface';
import { getToken } from '../../tokens/tokens.utils';
import { Affiliate } from './affiliate.config';

export class Yearn extends Affiliate {
  constructor() {
    super(Protocol.Yearn);
    Affiliate.register(Protocol.Yearn, this);
  }

  async getAffiliateVaultData(chain: Chain, sett: SettDefinition): Promise<SettAffiliateData> {
    const affiliateData: SettAffiliateData = {
      protocol: Protocol.Yearn,
    };
    if (sett.hasBouncer) {
      try {
        const vaultWrapper = new ethers.Contract(sett.settToken, yearnAffiliateVaultWrapperAbi, chain.provider);
        const guestListAddress = await vaultWrapper.guestList();
        const guesList = new ethers.Contract(guestListAddress, guestListAbi, chain.provider);
        const [available, limit]: [BigNumber, BigNumber] = await Promise.all([
          guesList.remainingTotalDepositAllowed(),
          guesList.totalDepositCap(),
        ]);
        const depositToken = getToken(sett.depositToken);
        affiliateData.availableDepositLimit = parseFloat(ethers.utils.formatUnits(available, depositToken.decimals));
        affiliateData.depositLimit = parseFloat(ethers.utils.formatUnits(limit, depositToken.decimals));
      } catch {
        affiliateData.availableDepositLimit = 0;
        affiliateData.depositLimit = 50;
      }
    }
    return affiliateData;
  }
}
