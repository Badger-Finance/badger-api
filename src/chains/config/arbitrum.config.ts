import { Protocol } from '../../config/enums/protocol.enum';
import { SettState } from '../../config/enums/sett-state.enum';
import { Stage } from '../../config/enums/stage.enum';
import rpc from '../../config/rpc.config';
import { TOKENS } from '../../config/tokens.config';
import { GasPrices } from '../../gas/interfaces/gas-prices.interface';
import { getCurveSettTokenBalance } from '../../protocols/strategies/convex.strategy';
import { BouncerType } from '../../rewards/enums/bouncer-type.enum';
import { SettDefinition } from '../../setts/interfaces/sett-definition.interface';
import { arbitrumTokensConfig } from '../../tokens/config/arbitrum-tokens.config';
import { ChainNetwork } from '../enums/chain-network.enum';
import { ArbitrumStrategy } from '../strategies/arbitrum.strategy';
import { Chain } from './chain.config';

export class Arbitrum extends Chain {
  constructor() {
    super(
      'Arbitrum',
      'arbitrum',
      '0xa4b1',
      ChainNetwork.Arbitrum,
      arbitrumTokensConfig,
      arbitrumSetts,
      rpc[ChainNetwork.Arbitrum],
      new ArbitrumStrategy(Object.keys(arbitrumTokensConfig)),
      2425847,
      '0x635EB2C39C75954bb53Ebc011BDC6AfAAcE115A6',
      '0x85E1cACAe9a63429394d68Db59E14af74143c61c',
    );
    Chain.register(this.network, this);
  }

  async getGasPrices(): Promise<GasPrices> {
    const gasPrice = 2;
    return {
      rapid: gasPrice,
      fast: gasPrice,
      standard: gasPrice,
      slow: gasPrice,
    };
  }
}

export const arbitrumSetts: SettDefinition[] = [
  {
    name: 'Sushiswap Wrapped Ether/Sushi Helper',
    settToken: TOKENS.BARB_SUSHI_WETH_SUSHI,
    depositToken: TOKENS.ARB_SUSHI_WETH_SUSHI,
    createdBlock: 13163959,
    protocol: Protocol.Sushiswap,
    strategy: '0x86f772C82914f5bFD168f99e208d0FC2C371e9C2',
  },
  {
    name: 'Sushiswap Wrapped BTC/Wrapped ETH',
    settToken: TOKENS.BARB_SUSHI_WETH_WBTC,
    depositToken: TOKENS.ARB_SUSHI_WETH_WBTC,
    createdBlock: 13163959,
    protocol: Protocol.Sushiswap,
    strategy: '0xA6827f0f14D0B83dB925B616d820434697328c22',
  },
  {
    name: 'Curve renBTC/wBTC',
    settToken: TOKENS.BARB_CRV_RENBTC,
    depositToken: TOKENS.ARB_CRV_RENBTC,
    getTokenBalance: getCurveSettTokenBalance,
    createdBlock: 13237551,
    protocol: Protocol.Curve,
  },
  {
    name: 'Curve Tricrypto',
    settToken: TOKENS.BARB_CRV_TRICRYPTO,
    depositToken: TOKENS.ARB_CRV_TRICRYPTO,
    getTokenBalance: getCurveSettTokenBalance,
    createdBlock: 13237551,
    protocol: Protocol.Curve,
  },
  {
    name: 'Curve Tricrypto Light',
    settToken: TOKENS.BARB_CRV_TRICRYPTO_LITE,
    depositToken: TOKENS.ARB_CRV_TRICRYPTO,
    getTokenBalance: getCurveSettTokenBalance,
    createdBlock: 13321375,
    stage: Stage.Staging,
    state: SettState.Experimental,
    protocol: Protocol.Curve,
  },
  {
    name: 'Swapr Swapr/Wrapped ETH',
    settToken: TOKENS.BARB_SWP_SWPR_WETH,
    depositToken: TOKENS.ARB_SWP_SWPR_WETH,
    createdBlock: 13315350,
    state: SettState.Guarded,
    protocol: Protocol.Swapr,
  },
  {
    name: 'Swapr Wrapped BTC/Wrapped ETH',
    settToken: TOKENS.BARB_SWP_WBTC_WETH,
    depositToken: TOKENS.ARB_SWP_WBTC_WETH,
    createdBlock: 13315350,
    state: SettState.Guarded,
    protocol: Protocol.Swapr,
  },
  {
    name: 'Swapr Badger/Wrapped ETH',
    bouncer: BouncerType.Badger,
    settToken: TOKENS.BARB_SWP_BADGER_WETH,
    depositToken: TOKENS.ARB_SWP_BADGER_WETH,
    createdBlock: 2188169,
    state: SettState.Guarded,
    protocol: Protocol.Swapr,
  },
];
