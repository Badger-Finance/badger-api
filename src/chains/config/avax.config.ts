import { Network, Protocol, VaultState, VaultVersion } from '@badger-dao/sdk';
import rpc from '../../config/rpc.config';
import { TOKENS } from '../../config/tokens.config';
import { GasPrices } from '../../gas/interfaces/gas-prices.interface';
import { VaultDefinition } from '../../vaults/interfaces/vault-definition.interface';
import { Chain } from './chain.config';
import { BaseStrategy } from '../strategies/base.strategy';
import { avalancheTokensConfig } from '../../tokens/config/avax-tokens.config';
import { Stage } from '../../config/enums/stage.enum';

export class Avalanche extends Chain {
  constructor() {
    super(
      'Avalanche',
      'avalanche',
      '0xa86a',
      Network.Avalanche,
      avalancheTokensConfig,
      rpc[Network.Avalanche],
      avalancheSetts,
      new BaseStrategy(Network.Avalanche, Object.keys(avalancheTokensConfig)),
    );
    Chain.register(this.network, this);
  }

  async getGasPrices(): Promise<GasPrices> {
    return this.defaultGasPrice();
  }

  // TODO: Update if badger is ever updated
  getBadgerTokenAddress(): string {
    return TOKENS.BADGER;
  }
}

export const avalancheSetts: VaultDefinition[] = [
  {
    name: 'WBTC',
    vaultToken: TOKENS.BAVAX_WBTC,
    depositToken: TOKENS.AVAX_WBTC,
    protocol: Protocol.Aave,
    stage: Stage.Staging,
    state: VaultState.Experimental,
    version: VaultVersion.v1_5,
  },
];
