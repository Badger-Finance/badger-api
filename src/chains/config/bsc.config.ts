import { Network } from "@badger-dao/sdk";

import rpc from "../../config/rpc.config";
import { TOKENS } from "../../config/tokens.config";
import { bscTokensConfig } from "../../tokens/config/bsc-tokens.config";
import { BaseStrategy } from "../strategies/base.strategy";
import { Chain } from "./chain.config";

export class BinanceSmartChain extends Chain {
  constructor() {
    super(Network.BinanceSmartChain, bscTokensConfig, rpc[Network.BinanceSmartChain], new BaseStrategy(Network.BinanceSmartChain));
    Chain.register(this.network, this);
  }

  getBadgerTokenAddress(): string {
    return TOKENS.MULTI_BADGER;
  }
}
