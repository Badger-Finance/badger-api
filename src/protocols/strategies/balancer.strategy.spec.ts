import { Erc20, Erc20__factory } from '@badger-dao/sdk';
import { BigNumber, ethers } from 'ethers';
import { mock, MockProxy } from 'jest-mock-extended';

import { CurrentVaultSnapshotModel } from '../../aws/models/current-vault-snapshot.model';
import { Chain } from '../../chains/config/chain.config';
import { TOKENS } from '../../config/tokens.config';
import {
  BalancerVault,
  BalancerVault__factory,
  StablePhantomVault,
  StablePhantomVault__factory,
  WeightedPool,
  WeightedPool__factory,
} from '../../contracts';
import * as balancerGraph from '../../graphql/generated/balancer';
import {
  MOCK_TOKEN,
  MOCK_TOKENS,
  MOCK_VAULT_DEFINITION,
  MOCK_VAULT_SNAPSHOT,
  TEST_ADDR,
  TEST_CURRENT_BLOCK,
  TEST_CURRENT_TIMESTAMP,
  TEST_TOKEN,
} from '../../test/constants';
import { setupMockChain } from '../../test/mocks.utils';
import * as vaultsUtils from '../../vaults/vaults.utils';
import {
  getBalancerPoolTokens,
  getBalancerVaultTokenBalance,
  getBalancerYieldSources,
  getBPTPrice,
  resolveBalancerPoolTokenPrice,
} from './balancer.strategy';

describe('balancer.strategy', () => {
  let chain: Chain;
  let stablePool: MockProxy<StablePhantomVault>;
  let weightedPool: MockProxy<WeightedPool>;

  beforeEach(() => {
    chain = setupMockChain();

    jest.spyOn(console, 'error').mockImplementation();
    jest.spyOn(balancerGraph, 'getSdk').mockImplementation((_client) => {
      return {
        PoolSnapshots(_v, _r): Promise<balancerGraph.PoolSnapshotsQuery> {
          return Promise.resolve({
            poolSnapshots: [
              {
                id: TEST_TOKEN,
                swapFees: 1500000,
                liquidity: 1000,
                amounts: [],
                timestamp: TEST_CURRENT_TIMESTAMP,
                totalShares: 100,
                swapVolume: 1000,
              },
              {
                id: TEST_TOKEN,
                swapFees: 1600000,
                liquidity: 1000,
                amounts: [],
                timestamp: TEST_CURRENT_TIMESTAMP,
                totalShares: 100,
                swapVolume: 1000,
              },
              {
                id: TEST_TOKEN,
                swapFees: 1700000,
                liquidity: 1000,
                amounts: [],
                timestamp: TEST_CURRENT_TIMESTAMP,
                totalShares: 100,
                swapVolume: 1000,
              },
            ],
          });
        },
      };
    });

    stablePool = mock<StablePhantomVault>();
    jest.spyOn(stablePool, 'getBptIndex').mockImplementation(async () => BigNumber.from(1));
    jest.spyOn(StablePhantomVault__factory, 'connect').mockImplementation(() => stablePool);

    weightedPool = mock<WeightedPool>();
    jest.spyOn(weightedPool, 'getVault').mockImplementation(async () => TEST_ADDR);
    jest.spyOn(weightedPool, 'getPoolId').mockImplementation(async () => TEST_ADDR);
    jest.spyOn(WeightedPool__factory, 'connect').mockImplementation(() => weightedPool);

    const balancerVault = mock<BalancerVault>();
    const poolTokens = <
      [string[], BigNumber[], BigNumber] & {
        tokens: string[];
        balances: BigNumber[];
        lastChangeBlock: BigNumber;
      }
    >{
      tokens: [TOKENS.BB_A_USDT, TOKENS.DIGG, TOKENS.BB_A_USDC, TOKENS.BB_A_DAI],
      balances: [
        ethers.constants.WeiPerEther,
        ethers.constants.WeiPerEther,
        ethers.constants.WeiPerEther,
        ethers.constants.WeiPerEther,
      ],
      lastChangeBlock: BigNumber.from(TEST_CURRENT_BLOCK),
    };
    // throw in digg as the ignored phantom token
    jest.spyOn(balancerVault, 'getPoolTokens').mockImplementation(async () => poolTokens);
    jest.spyOn(BalancerVault__factory, 'connect').mockImplementation(() => balancerVault);
  });

  describe('getBalancerYieldSources', () => {
    it('returns no yield sources if it encounters an error', async () => {
      jest.spyOn(balancerGraph, 'getSdk').mockImplementation((_client) => {
        throw new Error('Expected test error: getSdk');
      });
      const result = await getBalancerYieldSources(MOCK_VAULT_DEFINITION);
      expect(result).toMatchObject([]);
    });

    it('returns no yield sources if not enough data is available', async () => {
      jest.spyOn(balancerGraph, 'getSdk').mockImplementation((_client) => {
        return {
          PoolSnapshots(_v, _r): Promise<balancerGraph.PoolSnapshotsQuery> {
            return Promise.resolve({
              poolSnapshots: [],
            });
          },
        };
      });
      const result = await getBalancerYieldSources(MOCK_VAULT_DEFINITION);
      expect(result).toMatchObject([]);
    });

    it('returns fee data for the requested pool', async () => {
      const result = await getBalancerYieldSources(MOCK_VAULT_DEFINITION);
      expect(result).toMatchSnapshot();
    });
  });

  describe('getBalancerPoolTokens', () => {
    it('returns all relevant pool tokens', async () => {
      const result = await getBalancerPoolTokens(chain, TOKENS.BPT_BB_AAVE_USD);
      expect(result).toMatchSnapshot();
    });
  });

  describe('getBPTPrice', () => {
    it('calculates price for the phantom pool', async () => {
      jest.spyOn(stablePool, 'getVirtualSupply').mockImplementation(async () => ethers.constants.WeiPerEther);
      const result = await getBPTPrice(chain, TEST_TOKEN);
      expect(result).toMatchSnapshot();
    });

    it('calculates price for the weighted pool', async () => {
      jest.spyOn(stablePool, 'getVirtualSupply').mockImplementation(async () => {
        throw new Error('Expected test error: getVirtualSupply');
      });
      const erc20 = mock<Erc20>();
      jest.spyOn(erc20, 'totalSupply').mockImplementation(async () => ethers.constants.WeiPerEther);
      jest.spyOn(Erc20__factory, 'connect').mockImplementation(() => erc20);
      const result = await getBPTPrice(chain, TEST_TOKEN);
      expect(result).toMatchSnapshot();
    });
  });

  describe('getBalancerVaultTokenBalance', () => {
    beforeEach(() => {
      jest
        .spyOn(vaultsUtils, 'getCachedVault')
        .mockImplementation(async (_c, _v) => MOCK_VAULT_SNAPSHOT as CurrentVaultSnapshotModel);
    });

    it('returns vault tokens for a stable pool', async () => {
      jest.spyOn(stablePool, 'getVirtualSupply').mockImplementation(async () => ethers.constants.WeiPerEther);
      const result = await getBalancerVaultTokenBalance(chain, TEST_TOKEN);
      expect(result).toMatchSnapshot();
    });

    it('returns vault tokens for a weighted pool', async () => {
      jest.spyOn(stablePool, 'getVirtualSupply').mockImplementation(async () => {
        throw new Error('Expected test error: getVirtualSupply');
      });
      const erc20 = mock<Erc20>();
      jest.spyOn(erc20, 'totalSupply').mockImplementation(async () => ethers.constants.WeiPerEther);
      jest.spyOn(Erc20__factory, 'connect').mockImplementation(() => erc20);
      const result = await getBalancerVaultTokenBalance(chain, TEST_TOKEN);
      expect(result).toMatchSnapshot();
    });
  });

  describe('resolveBalancerPoolTokenPrice', () => {
    it('returns zero price when encountering errors', async () => {
      jest.spyOn(weightedPool, 'getNormalizedWeights').mockImplementation(async () => {
        throw new Error('Expected test error: getNormalizedWeights');
      });
      const { price } = await resolveBalancerPoolTokenPrice(chain, MOCK_TOKEN, TEST_ADDR);
      expect(price).toEqual(0);
    });

    it('returns zero price for missing requested token', async () => {
      jest
        .spyOn(weightedPool, 'getNormalizedWeights')
        .mockImplementation(async () => [
          ethers.constants.WeiPerEther.mul(30),
          ethers.constants.WeiPerEther.mul(40),
          ethers.constants.WeiPerEther.mul(30),
        ]);
      const { price } = await resolveBalancerPoolTokenPrice(chain, MOCK_TOKEN, TEST_ADDR);
      expect(price).toEqual(0);
    });

    it('returns the price of a token in a weighted pool', async () => {
      jest
        .spyOn(weightedPool, 'getNormalizedWeights')
        .mockImplementation(async () => [
          ethers.constants.WeiPerEther.mul(30),
          ethers.constants.WeiPerEther.mul(40),
          ethers.constants.WeiPerEther.mul(30),
        ]);
      const token = MOCK_TOKENS[TOKENS.BB_A_USDT];
      const result = await resolveBalancerPoolTokenPrice(chain, token, TEST_ADDR);
      expect(result).toMatchSnapshot();
    });

    it('returns the price of a token in a stable pool', async () => {
      jest.spyOn(weightedPool, 'getNormalizedWeights').mockImplementation(async () => {
        throw new Error('Expected test error: getNormalizedWeights');
      });
      const amplificationData = <
        [BigNumber, boolean, BigNumber] & {
          value: BigNumber;
          isUpdating: boolean;
          precision: BigNumber;
        }
      >{
        value: ethers.constants.WeiPerEther,
        isUpdating: false,
        precision: BigNumber.from('1000000'),
      };
      jest.spyOn(stablePool, 'getAmplificationParameter').mockImplementation(async () => amplificationData);
      const invariantData = <
        [BigNumber, BigNumber] & {
          lastInvariant: BigNumber;
          lastInvariantAmp: BigNumber;
        }
      >{
        lastInvariant: ethers.constants.WeiPerEther,
        lastInvariantAmp: BigNumber.from('1000000'),
      };
      jest.spyOn(stablePool, 'getLastInvariant').mockImplementation(async () => invariantData);

      const balancerVault = mock<BalancerVault>();
      const poolTokens = <
        [string[], BigNumber[], BigNumber] & {
          tokens: string[];
          balances: BigNumber[];
          lastChangeBlock: BigNumber;
        }
      >{
        tokens: [TOKENS.BB_A_USDT, TOKENS.BB_A_USDC],
        balances: [ethers.constants.WeiPerEther, ethers.constants.WeiPerEther],
        lastChangeBlock: BigNumber.from(TEST_CURRENT_BLOCK),
      };
      // throw in digg as the ignored phantom token
      jest.spyOn(balancerVault, 'getPoolTokens').mockImplementation(async () => poolTokens);
      jest.spyOn(BalancerVault__factory, 'connect').mockImplementation(() => balancerVault);

      const token = MOCK_TOKENS[TOKENS.BB_A_USDT];
      const result = await resolveBalancerPoolTokenPrice(chain, token, TEST_ADDR);
      expect(result).toMatchSnapshot();
    });
  });
});
