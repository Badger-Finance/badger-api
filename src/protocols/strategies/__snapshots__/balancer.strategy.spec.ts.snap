// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`balancer.strategy getBPTPrice calculates price for the phantom pool 1`] = `
Object {
  "address": "0x3472A5A71965499acd81997a54BBA8D852C6E53d",
  "price": 81435,
}
`;

exports[`balancer.strategy getBPTPrice calculates price for the weighted pool 1`] = `
Object {
  "address": "0x3472A5A71965499acd81997a54BBA8D852C6E53d",
  "price": 81435,
}
`;

exports[`balancer.strategy getBalancerPoolTokens returns all relevant pool tokens 1`] = `
Array [
  Object {
    "address": "0x2BBf681cC4eb09218BEe85EA2a5d3D13Fa40fC0C",
    "balance": 1,
    "decimals": 18,
    "name": "Balancer Aave Boosted Pool (USDT)",
    "symbol": "bb-a-USDT",
    "value": 11199,
  },
  Object {
    "address": "0x9210F1204b5a24742Eba12f710636D76240dF3d0",
    "balance": 1,
    "decimals": 18,
    "name": "Balancer Aave Boosted Pool (USDC)",
    "symbol": "bb-a-USDC",
    "value": 37392,
  },
  Object {
    "address": "0x804CdB9116a10bB78768D3252355a1b18067bF8f",
    "balance": 1,
    "decimals": 18,
    "name": "Balancer Aave Boosted Pool (DAI)",
    "symbol": "bb-a-DAI",
    "value": 32844,
  },
]
`;

exports[`balancer.strategy getBalancerVaultTokenBalance returns vault tokens for a stable pool 1`] = `
VaultTokenBalance {
  "chain": "ethereum",
  "id": "ethereum-0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
  "tokenBalances": Array [
    Object {
      "address": "0x2BBf681cC4eb09218BEe85EA2a5d3D13Fa40fC0C",
      "balance": 48405.45345505732,
      "decimals": 18,
      "name": "Balancer Aave Boosted Pool (USDT)",
      "symbol": "bb-a-USDT",
      "value": 542092673.243187,
    },
    Object {
      "address": "0x9210F1204b5a24742Eba12f710636D76240dF3d0",
      "balance": 48405.45345505732,
      "decimals": 18,
      "name": "Balancer Aave Boosted Pool (USDC)",
      "symbol": "bb-a-USDC",
      "value": 1809976715.5915034,
    },
    Object {
      "address": "0x804CdB9116a10bB78768D3252355a1b18067bF8f",
      "balance": 48405.45345505732,
      "decimals": 18,
      "name": "Balancer Aave Boosted Pool (DAI)",
      "symbol": "bb-a-DAI",
      "value": 1589828713.2779026,
    },
  ],
  "vault": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
}
`;

exports[`balancer.strategy getBalancerVaultTokenBalance returns vault tokens for a weighted pool 1`] = `
VaultTokenBalance {
  "chain": "ethereum",
  "id": "ethereum-0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
  "tokenBalances": Array [
    Object {
      "address": "0x2BBf681cC4eb09218BEe85EA2a5d3D13Fa40fC0C",
      "balance": 48405.45345505732,
      "decimals": 18,
      "name": "Balancer Aave Boosted Pool (USDT)",
      "symbol": "bb-a-USDT",
      "value": 542092673.243187,
    },
    Object {
      "address": "0x9210F1204b5a24742Eba12f710636D76240dF3d0",
      "balance": 48405.45345505732,
      "decimals": 18,
      "name": "Balancer Aave Boosted Pool (USDC)",
      "symbol": "bb-a-USDC",
      "value": 1809976715.5915034,
    },
    Object {
      "address": "0x804CdB9116a10bB78768D3252355a1b18067bF8f",
      "balance": 48405.45345505732,
      "decimals": 18,
      "name": "Balancer Aave Boosted Pool (DAI)",
      "symbol": "bb-a-DAI",
      "value": 1589828713.2779026,
    },
  ],
  "vault": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
}
`;

exports[`balancer.strategy getBalancerYieldSources returns fee data for the requested pool 1`] = `
Array [
  CachedYieldSource {
    "address": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    "boostable": false,
    "chain": "ethereum",
    "chainAddress": "ethereum-0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    "id": "ethereum-0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28_lp_fee_balancer_lp_fees_flat",
    "name": "Balancer LP Fees",
    "performance": Object {
      "baseYield": 1825000,
      "grossYield": 1825000,
      "maxGrossYield": 1825000,
      "maxYield": 1825000,
      "minGrossYield": 1825000,
      "minYield": 1825000,
    },
    "type": "lp_fee",
  },
]
`;

exports[`balancer.strategy resolveBalancerPoolTokenPrice returns the price of a token in a stable pool 1`] = `
Object {
  "address": "0x2BBf681cC4eb09218BEe85EA2a5d3D13Fa40fC0C",
  "price": 0,
}
`;

exports[`balancer.strategy resolveBalancerPoolTokenPrice returns the price of a token in a weighted pool 1`] = `
Object {
  "address": "0x2BBf681cC4eb09218BEe85EA2a5d3D13Fa40fC0C",
  "price": 70236,
}
`;
